### PR TITLE
Add MTA buses config

### PR DESCRIPTION
**Pull request by cedarbaum**
# About

This change adds static and realtime trip data for MTA buses.

- The realtime feed was configured base on: http://bt.mta.info/wiki/Developers/GTFSRt
- The static feeds were configured based on: http://web.mta.info/developers/developer-data-terms.html#data

# Issue with trip loops

In my testing so far, this appears to work mostly out of the box except for one error that sometime occurs for routes:

```
service map cannot be topologically sorted: the provided graph is not topologically sortable because it contains a cycle
```

I looked into this a bit, and it seems there are loops being created in some GTFS trips. For example, on the [M96](https://bustime.mta.info/m/?q=M96) route, I observed this:

image

My guess would be this is a bug with the feed, but it's hard to say for sure without a repro outside of transiter. I did try writing a small Python script to poll the feed and look for loops within trips, but I haven't yet seen the same behavior there (this could just be an issue of timing though).

It does seem transiter is resilient to this sort of error, though I am curious if you think it's something we need to understand better before using this config. Also, I am curious if you have any thoughts about where the bug most likely is and what could be done to further investigate.

# Other notes:

1. The GTFS realtime feeds appear to work even without an API key specified in the URL. The official documentation still says to use one, so it is probably still best to follow that.
2. The documented GTFS realtime feed URLs use HTTP for transport, but HTTPS does work so I've used that instead.

As always, I appreciate any time you're able to spend looking over this and please let me know if you have any feedback when you have a chance.

### CODE REVIEW

**Comment by cedarbaum** [comment]
Quick follow-up - I believe I now understand the trip loop issue. Apparently trip ids are not unique for MTA bus trips. This causes transiter to try and reconcile the stop sequences of multiple trips running on the same route, which creates the cycles.

I believe creating an id from a combination of `TripId` and `VehicleId` will be unique, so this could be done as an extension in the `gtfs` library. I'll look into this next.

**Comment by cedarbaum** [comment]
Thanks for the initial feedback, and it would be great to have you try it on your end when you are back and have time! In the meantime, I've done some more experimentation and have some additional notes below:

1. As mentioned previously, `TripId`s are not unique. I believe this can be fixed (at least in the majority of cases) with this PR: https://github.com/jamespfennell/gtfs/pull/2
2. In addition to non-unique `TripId`s, stop sequences also don't appear to be stable across updates (**though I want to double-check my work here, since this seems very strange, even for GTFS 🙂**). For example, consider the following 2 updates for a trip:

### Update 1
```
[{'stopSequence': 0, 'arrival': {'time': '1679544559'}, 'departure': {'time': '1679544559'}, 'stopId': '305733'},
 {'stopSequence': 1, 'arrival': {'time': '1679544574'}, 'departure': {'time': '1679544574'}, 'stopId': '305734'},
 {'stopSequence': 2, 'arrival': {'time': '1679544607'}, 'departure': {'time': '1679544607'}, 'stopId': '305735'}]
```

### Update 2
```
[{'stopSequence': 2, 'arrival': {'time': '1679544303'}, 'departure': {'time': '1679544303'}, 'stopId': '305734'},
 {'stopSequence': 3, 'arrival': {'time': '1679544336'}, 'departure': {'time': '1679544336'}, 'stopId': '305735'},
 {'stopSequence': 4, 'arrival': {'time': '1679544376'}, 'departure': {'time': '1679544376'}, 'stopId': '305736'}],
```

This causes some issues when transiter tries to merge the trip timeliness across updates, most noticeably the loop issue mentioned previously.

I was able to fix this by simply deleting all previous stop times on trip updates (i.e. MTA bus trips will only return current and futures stops now). Something more clever could probably be done to merge the timelines.

**Comment by jamespfennell** [comment]
Wow this system has over 12 thousand stops!

**Comment by jamespfennell** [comment]
Hmmm just as an FYI, I'm getting lots of "service map cannot be topologically sorted" errors but no duplicate trip IDs. So perhaps there is another reason these maps can't be sorted, in addition to the trip ID duplication issue? I'm going to look further.

**Comment by cedarbaum** [comment]
Hey @jamespfennell - did you see my theory about unstable stop sequence numbers in my last comment? It may explain why you're still seeing that error. When I completely replace all stop sequences with each new update (i.e. make no attempt to preserve trip history), I no longer get the error. I also validated this theory with a python script that compares stop sequences between updates for the same trip id. Let me know if that makes sense based on what you're seeing on your end.

**Comment by jamespfennell** [comment]
Ah yes, sorry I did read that comment last week but had forgotten it. Do you have any opinion on the best way to fix this? Should Transiter ignore the stop sequences numbers and generate its own? This is what it does for the subway.

Though I do wonder - is the MTA using stop sequences because some trips call at the same stop twice? In this case the GTFS realtime spec does require stop sequences be provided to disambiguate the stop times for the same stop.

**Comment by jamespfennell** [comment]
By the way, I'm happy to merge in any version of this work that you're happy with. Things like improving the stop sequences situation can always be done in subsequent PRs (or not at all!). From my perspective, even just the new config file (along with the `transfers.txt` fix) is a great addition. Totally up to you.

**Comment by jamespfennell** [comment]
I think there is a quick fix for the service maps issue alone. When calculating the service map, we could skip all stops times that are in the past. It's a one line fix to [this one SQL query](https://github.com/jamespfennell/transiter/blob/29e282f1453294e9eb78d2a4aeab1139124f2c8d/db/queries/servicemap_queries.sql#L34) - it just needs an additional `WHERE trip_stop_time.future IS TRUE` clause. I should probably do this anyway it seems? Like the realtime service maps should show only current/future service

This wouldn't fix the trip data itself, and as in your comment above, we would still see duplicate past stops.

**Comment by cedarbaum** [comment]
Thanks for the service maps fix! I agree that this can be merged safely with just that (after bumping `gtfs` version and also addressing other comments). However, I also think there is some value in either/or of the below configuration options (which could be added to the YAML definitions):

1.) `removeTripHistory`: this option, when true, would make no attempt to keep past trips. This would probably be fine for many use cases (e.g., arrival boards).
2.) `reassignStopSequences`: this option (based on your suggestion) would ignore the stop sequences from GTFS and assign sequentially (assuming no duplicate stop ids). I _think_ this would be OK in most cases but, as you mentioned above, it would fail for routes that do indeed revisit stops.

So going forward, I can cleanup this PR and potentially also work on (1) or (2) if you think this an OK approach. Let me know your thoughts!

**Comment by jamespfennell** [comment]
Awesome, that sounds great!

I personally think (2) is the way to go because I _think_ it will end up making the MTA buses work correctly with trip history. The implementation shouldn't be too bad either? After adding the `reassignStopSequences` field to the feed config, I think it would just be a code change in [realtime.go](https://github.com/jamespfennell/transiter/blob/0bd406b35ce28160f956f10b4c3d181df0dd7384/internal/update/realtime/realtime.go#L42) to clear the stop sequence numbers if the new config field is true. This clearing just needs to happen before the stop sequences are automatically assigned by Transiter.

If you wanted to implement that that would be awesome! Could do it in this PR, or a follow up PR.

---

By the way, I think the downside of the new configuration option only really applies if you're trying to link the realtime data with the static data. My understanding is that the stop sequences give you a way of linking a stop time prediction in the realtime data to the associated scheduled time in GTFS static. For the MTA buses this realtime-static linking will presumably always be broken anyway because of the bug you discovered. But it's definitely a limitation of the new option that we should document. (To be clear Transiter currently doesn't attempt to do this realtime-static linking, but it's something I want to do in the future because it's needed to get some transit systems working, like the SF BART.)

---

Another random interesting thing (sorry for the brain dump) is that the MTA bus transit system is really big, and when I play with it locally I'm seeing trip feed updates take 7 or 8 seconds. Having this system in the repo will be great for performance benchmarking. I just took a pprof profile out of curiosity, and am finding that a lot of the slowness is because Transiter updates trip stop times with individual SQL queries, one at a time. This is slow just because for each SQL query, a network call has to be made to Postgres. Maybe in the future there is some optimization that could be done based on this kind of profiling.

**Comment by cedarbaum** [comment]
> Another random interesting thing (sorry for the brain dump) is that the MTA bus transit system is really big, and when I play with it locally I'm seeing trip feed updates take 7 or 8 seconds. Having this system in the repo will be great for performance benchmarking. I just took a pprof profile out of curiosity, and am finding that a lot of the slowness is because Transiter updates trip stop times with individual SQL queries, one at a time. This is slow just because for each SQL query, a network call has to be made to Postgres. Maybe in the future there is some optimization that could be done based on this kind of profiling.

Interesting! I was sort of thinking about this a bit when I was looking at that code, but figured Postgres would do some magic to optimize these operations within the same transaction. Do you think a multi-row `UPSERT` would help? Or, in the case of my hack to forget trip history, a full delete + multi-row `INSERT`.

**Comment by jamespfennell** [comment]
Yeah I was thinking a multi-row upsert (or even just update because inserts are relatively rare) would befaster. The main motivation would be to avoid the overhead of making a SQL call to Postgres for each stop time update update. Not sure how hard it is though! I think [Postgres supports it](https://stackoverflow.com/questions/18797608/update-multiple-rows-in-same-query-using-postgresql), but not sure about `sqlc`.

**Comment by jamespfennell** [comment]
I created a new issue for the optimizing the feed updater, just so that I don't derail your PR :) https://github.com/jamespfennell/transiter/issues/100

File: systems/us-ny-nycbus.yaml [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+  - id: "alerts"
+    updateStrategy: PERIODIC
+    updatePeriodS: 5
+
+    url: "http://gtfsrt.prod.obanyc.com/alerts?key={{ .Args.mta_bus_api_key }}"
+
+    parser: GTFS_REALTIME
+    gtfsRealtimeOptions:
+        extension: NYCT_BUS_TRIPS
+        reassignStopSequences: true
```
Comment (jamespfennell) [inline systems/us-ny-nycbus.yaml:56]: Looks like this is on the wrong feed config - should be trips above?

Comment (cedarbaum) [reply systems/us-ny-nycbus.yaml:56]: Ah, that's embarrassing, good catch! Will fix and send an update soon.

**Comment by jamespfennell** [comment]
Awesome, thank you!

**Pull request merged**